jax/_src/lax/control_flow/loops.py

```diff
@@ -2020,6 +2022,97 @@ def map(f, xs):
   _, ys = scan(g, (), xs)
   return ys
 
+
+def _tree_map_multi_output(f, *args):
+  """Like tree_map, but for functions that return tuples."""
+  leaves, treedefs = zip(*builtins.map(tree_flatten, args))
+  if any(treedef != treedefs[0] for treedef in treedefs):
+    raise ValueError(f"argument treedefs do not match {treedefs=}")
+  outputs = zip(*builtins.map(f, *leaves))
+  return tuple(tree_unflatten(treedefs[0], out) for out in outputs)
+
+
+def _lax_map(f, *xs):
+  """Like lax.map, but supports multiple arguments like the built-in map."""
+  g = lambda _, x: ((), f(*x))
+  _, ys = jax.lax.scan(g, (), xs)
+  return ys
+
+
+def batch_vmap(
+  f: F,
+  in_axes: int | None | Sequence[Any] = 0,
+  out_axes: Any = 0,
+  *,
+  batch_size: int,
+) -> F:
+  """jax.vmap, but looping when the batch dimension exceeds batch_size.
+
+  Wrapping f with jax.remat() is recommended if you want to use batch_vmap
```

Comment:
What's the behavior when the vmap dimension is not divisible by `batch_size` ? Is it an error ? Is the remainder padded (hopefully not) ? ...

It would be nice to document this

Reply:
New implementation will yield an error.

Reply:
1. In our use-case, there's no guarantee `dim % batch_size == 0`. How difficult would it be to support arbitrary dimension ?
2. Otherwise, the limitation should be documented in the docstring of the function (this doesn't seems to be the case unless I missed something)

Reply:
Adding on to this, allowing a batch size that doesn't evenly divide the dimension would be very useful. I had to solve this problem in my own work, and ended up just running the divisible portion via scan and the remainder separately. While that might not be a perfect solution since it's two separate calls, it at least allows seamless usage:

```python
    def batched_vmap(f, batch_size: int, axis: int = 0):
        vmap_f = jax.vmap(f)
        scan_f = lambda carry, x: (None, vmap_f(x))

        def wrapper(array):
            # Separate into [non-]divisible parts, and reshape for scan
            array = jnp.moveaxis(array, axis, 0)
            batch_count = len(array) // batch_size
            array,extra = jnp.split(array, [batch_count * batch_size])
            array = array.reshape((batch_count, batch_size) + array.shape[1:])

            # Scan over the divisible section, and separately apply vmap to extra
            array = jax.lax.scan(scan_f, None, array)[1]
            extra = vmap_f(extra)

            # Recombine the two parts, and restore the axis position
            array = array.reshape((-1,) + array.shape[2:])
            array = jnp.append(array, extra, axis=0)
            return jnp.moveaxis(array, 0, axis)
        return wrapper
```

I'm certainly not an expert with jax, though - so I might be missing some reasons why that approach shouldn't be used.